src/RoomDirectoryModel.cpp

```diff
@@ -0,0 +1,74 @@
+// SPDX-FileCopyrightText: 2021 Nheko Contributors
+//
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "RoomDirectoryModel.h"
+#include "CompletionModelRoles.h"
+
+RoomDirectoryModel::RoomDirectoryModel(std::string &s, QObject *parent)
+                    : QAbstractListModel(parent)
+                    , server_(s)
+{
+    // ??? don't need to do anything special here ???
+}
+
+QHash<int, QByteArray> 
+RoomDirectoryModel::roleNames() const
+{
+    // ??? TODO once I understand what the roles actually mean
```

Comment:
Roles are what properties you want to have available for every element in the list. For example: room name, avatar url, member count, network, etc. Those can be accessed via the role name from Qml, by doing model.<rolename> in the delegate definition. See this for an example: https://doc.qt.io/qt-5/qtquick-modelviewsdata-cppmodels.html#qabstractitemmodel-subclass (The required properties are a new thing in Qt5.14, we can't use them yet, maybe check an older version of those docs.